// @dexwox-labs/a2a-core/dist/conversion/type-converters.js
"use strict";
/**
 * @module TypeConverters
 * @description Utility functions for converting between A2A protocol types and JSON
 *
 * This module provides functions for converting between A2A protocol types and
 * their JSON representations, as well as between A2A protocol types and JSON-RPC
 * request/response types.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.toJsonRpcRequest = toJsonRpcRequest;
exports.fromJsonRpcResponse = fromJsonRpcResponse;
exports.agentCardToJson = agentCardToJson;
exports.jsonToAgentCard = jsonToAgentCard;
exports.taskToJson = taskToJson;
exports.jsonToTask = jsonToTask;
/**
 * Converts an A2A protocol discover request to a JSON-RPC request
 *
 * This function transforms an A2A protocol discover request into a standard
 * JSON-RPC 2.0 request format, ensuring compatibility with JSON-RPC servers.
 *
 * @param request - The A2A protocol discover request to convert
 * @returns A JSON-RPC 2.0 formatted request
 *
 * @example
 * ```typescript
 * const discoverRequest: DiscoverRequest = {
 *   id: '123',
 *   method: 'discoverAgents',
 *   params: { capability: 'chat' }
 * };
 *
 * const jsonRpcRequest = toJsonRpcRequest(discoverRequest);
 * // {
 * //   jsonrpc: '2.0',
 * //   id: '123',
 * //   method: 'discoverAgents',
 * //   params: { capability: 'chat' }
 * // }
 * ```
 */
function toJsonRpcRequest(request) {
    return {
        jsonrpc: '2.0',
        id: request.id,
        method: request.method,
        params: request.params
    };
}
/**
 * Extracts the result from a JSON-RPC response
 *
 * This function extracts the result from a JSON-RPC response, throwing an error
 * if the response contains an error. It's a type-safe way to handle JSON-RPC
 * responses in the A2A protocol.
 *
 * @template T - The expected type of the result
 * @param response - The JSON-RPC response to extract the result from
 * @returns The result from the JSON-RPC response, cast to type T
 * @throws Error if the response contains an error
 *
 * @example
 * ```typescript
 * try {
 *   // Extract agents from a discover response
 *   const agents = fromJsonRpcResponse<AgentCard[]>(response);
 *   console.log(`Found ${agents.length} agents`);
 * } catch (error) {
 *   console.error('Error in JSON-RPC response:', error.message);
 * }
 * ```
 */
function fromJsonRpcResponse(response) {
    if (response.error) {
        throw new Error(response.error.message);
    }
    return response.result;
}
/**
 * Converts an AgentCard to a JSON-serializable object
 *
 * This function converts an AgentCard instance to a plain JavaScript object
 * that can be safely serialized to JSON. It creates a deep copy of the agent's
 * capabilities array to prevent modifications to the original object.
 *
 * @param agent - The AgentCard to convert
 * @returns A JSON-serializable representation of the agent
 *
 * @example
 * ```typescript
 * const agent: AgentCard = {
 *   id: 'assistant-agent',
 *   name: 'Assistant',
 *   capabilities: ['chat', 'answer-questions'],
 *   endpoint: 'https://example.com/agents/assistant'
 * };
 *
 * const json = agentCardToJson(agent);
 * const serialized = JSON.stringify(json);
 * ```
 */
function agentCardToJson(agent) {
    return {
        ...agent,
        capabilities: [...agent.capabilities]
    };
}
/**
 * Converts a JSON object to an AgentCard
 *
 * This function converts a plain JavaScript object (typically parsed from JSON)
 * into an AgentCard instance. It validates that the required fields are present
 * and ensures the capabilities array is properly formatted.
 *
 * @param json - The JSON object to convert
 * @returns An AgentCard instance
 *
 * @example
 * ```typescript
 * // Parse JSON from an API response
 * const responseData = JSON.parse(responseText);
 *
 * // Convert to an AgentCard
 * const agent = jsonToAgentCard(responseData);
 * console.log(`Agent: ${agent.name} (${agent.id})`);
 * console.log(`Capabilities: ${agent.capabilities.join(', ')}`);
 * ```
 */
function jsonToAgentCard(json) {
    return {
        id: json.id,
        name: json.name,
        capabilities: Array.isArray(json.capabilities) ? [...json.capabilities] : [],
        endpoint: json.endpoint
    };
}
/**
 * Converts a Task to a JSON-serializable object
 *
 * This function converts a Task instance to a plain JavaScript object that can
 * be safely serialized to JSON. It creates deep copies of the task's input and
 * output schemas to prevent modifications to the original object.
 *
 * @param task - The Task to convert
 * @returns A JSON-serializable representation of the task
 *
 * @example
 * ```typescript
 * const task: Task = {
 *   id: 'task-123',
 *   name: 'Answer Question',
 *   status: 'completed',
 *   createdAt: new Date().toISOString(),
 *   updatedAt: new Date().toISOString(),
 *   inputSchema: { type: 'object', properties: { question: { type: 'string' } } },
 *   outputSchema: { type: 'object', properties: { answer: { type: 'string' } } }
 * };
 *
 * const json = taskToJson(task);
 * const serialized = JSON.stringify(json);
 * ```
 */
function taskToJson(task) {
    return {
        ...task,
        inputSchema: task.inputSchema ? { ...task.inputSchema } : undefined,
        outputSchema: task.outputSchema ? { ...task.outputSchema } : undefined
    };
}
/**
 * Converts a JSON object to a Task
 *
 * This function converts a plain JavaScript object (typically parsed from JSON)
 * into a Task instance. It validates that the required fields are present and
 * sets default values for missing fields. By default, new tasks are created with
 * a 'submitted' status.
 *
 * @param json - The JSON object to convert
 * @returns A Task instance
 *
 * @example
 * ```typescript
 * // Parse JSON from an API response
 * const responseData = JSON.parse(responseText);
 *
 * // Convert to a Task
 * const task = jsonToTask(responseData);
 * console.log(`Task: ${task.name} (${task.id})`);
 * console.log(`Status: ${task.status}`);
 * console.log(`Created: ${new Date(task.createdAt).toLocaleString()}`);
 * ```
 */
function jsonToTask(json) {
    return {
        id: json.id,
        name: json.name,
        description: json.description,
        status: 'submitted',
        inputSchema: json.inputSchema,
        outputSchema: json.outputSchema,
        createdAt: json.createdAt || new Date().toISOString(),
        updatedAt: json.updatedAt || new Date().toISOString()
    };
}
